Code/EnginePlugins/OpenXRPlugin/OpenXRInputDevice.cpp

```diff
@@ -0,0 +1,348 @@
+  profileBindings.interactionProfile = InteractionProfile;
+  profileBindings.suggestedBindings = xrBindings.GetData();
+  profileBindings.countSuggestedBindings = xrBindings.GetCount();
+  EZ_CHECK_XR(xrSuggestInteractionProfileBindings(instance, &profileBindings));
+  return EZ_SUCCESS;
+}
+
+XrResult ezOpenXRInputDevice::AttachSessionActionSets(XrSession session)
+{
+  m_session = session;
+  XrSessionActionSetsAttachInfo attachInfo{ XR_TYPE_SESSION_ACTION_SETS_ATTACH_INFO };
+  ezHybridArray<XrActionSet, 1> actionSets;
+  actionSets.PushBack(m_ActionSet);
+
+  attachInfo.countActionSets = actionSets.GetCount();
+  attachInfo.actionSets = actionSets.GetData();
+  return xrAttachSessionActionSets(session, &attachInfo);
+}
+
+void ezOpenXRInputDevice::InitializeDevice()
+{
+}
+
+void ezOpenXRInputDevice::RegisterInputSlots()
+{
+  for (const Action& action : m_booleanActions)
+  {
+    RegisterInputSlot(action.sLeftKey, action.sLeftKey, ezInputSlotFlags::IsButton);
+    RegisterInputSlot(action.sRightKey, action.sRightKey, ezInputSlotFlags::IsButton);
+  }
+}
+
+void ezOpenXRInputDevice::UpdateActions()
+{
+  if (m_session == XR_NULL_HANDLE)
+    return;
+
+  EZ_PROFILE_SCOPE("UpdateInputSlotValues");
+  const XrFrameState& frameState = m_pOpenXR->m_frameState;
+
+  ezHybridArray<XrActiveActionSet, 1> activeActionSets;
+  activeActionSets.PushBack({m_ActionSet, XR_NULL_PATH });
+
+  XrActionsSyncInfo syncInfo{ XR_TYPE_ACTIONS_SYNC_INFO };
+  syncInfo.countActiveActionSets = activeActionSets.GetCount();
+  syncInfo.activeActionSets = activeActionSets.GetData();
+  XrResult res = xrSyncActions(m_session, &syncInfo);
+  if (res == XR_SESSION_NOT_FOCUSED)
+    return;
+
+  XR_CHECK_LOG(res);
+
+  for (const Action& action : m_booleanActions)
+  {
+    XrActionStateBoolean state{ XR_TYPE_ACTION_STATE_BOOLEAN };
+    XrActionStateGetInfo getInfo{ XR_TYPE_ACTION_STATE_GET_INFO };
+    getInfo.action = action.action;
+    getInfo.subactionPath = m_subActionPath[0];
+    XR_CHECK_LOG(xrGetActionStateBoolean(m_session, &getInfo, &state));
+    m_InputSlotValues[action.sLeftKey] = state.currentState;
+
+    getInfo.subactionPath = m_subActionPath[1];
+    XR_CHECK_LOG(xrGetActionStateBoolean(m_session, &getInfo, &state));
+    m_InputSlotValues[action.sRightKey] = state.currentState;
+  }
+
+  auto UpdatePose = [](ezVec3& vPosition, ezQuat& qRotation, bool& m_bIsValid, const XrSpaceLocation& viewInScene)
+  {
+    if ((viewInScene.locationFlags & (XR_SPACE_LOCATION_POSITION_VALID_BIT | XR_SPACE_LOCATION_ORIENTATION_VALID_BIT)) == (XR_SPACE_LOCATION_POSITION_VALID_BIT | XR_SPACE_LOCATION_ORIENTATION_VALID_BIT))
+    {
+      vPosition = ezOpenXR::ConvertPosition(viewInScene.pose.position);
+      qRotation = ezOpenXR::ConvertOrientation(viewInScene.pose.orientation);
+      m_bIsValid = true;
+    }
+    else
+    {
+      m_bIsValid = false;
+    }
+  };
+
+  for (ezUInt32 uiSide : {0, 1})
+  {
+    ezXRDeviceState& state = m_DeviceState[uiSide + 1];
```

Comment:
why +1 ?

Reply:
0 is head, 1 is left hand, 2 is right hand. Replaced the code with a more readable version.